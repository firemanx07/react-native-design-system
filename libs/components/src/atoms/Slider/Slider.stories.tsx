import { useNamespacedTheme } from '@proxym/themes';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen, StorybookSection } from '../../storybook';
import { default as SliderBase, SliderProps, SliderSize } from './Slider';

const Story: ComponentMeta<typeof SliderBase> = {
  component: SliderBase,
  title: 'Atoms/Slider',
  argTypes: {
    size: {
      control: { type: 'select' },
      options: Object.values(SliderSize).filter(
        item => typeof item === 'number',
      ),
    },
  },
};
export default Story;

const noop = () => {};

const Template: ComponentStory<typeof SliderBase> = (args: SliderProps) => {
  const { colors } = useNamespacedTheme();
  return (
    <StorybookScreen
      title="Slider:"
      style={{ backgroundColor: colors.graySoft }}
    >
      <StorybookSection width={(args.width ?? 370) + 30}>
        <SliderBase {...args} />
      </StorybookSection>
    </StorybookScreen>
  );
};

const parameters = {
  controls: {
    include: ['width', 'size', 'disabled', 'min', 'max', 'step', 'value'],
  },
};

const defaultArgs: SliderProps = {
  value: 0,
  min: 0,
  max: 100,
  onChange: noop,
  width: 300,
};

export const Slider = Template.bind({});
Slider.parameters = parameters;
Slider.args = {
  ...defaultArgs,
};
