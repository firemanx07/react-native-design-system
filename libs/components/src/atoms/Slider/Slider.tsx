import { styled } from '@proxym/themes';
import React from 'react';
import {
  PanGestureHandler,
  PanGestureHandlerGestureEvent,
} from 'react-native-gesture-handler';
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';

const BASE_MARKER_SIZE = 15;
export enum SliderSize {
  XL = BASE_MARKER_SIZE * 4,
  L = BASE_MARKER_SIZE * 3,
  M = BASE_MARKER_SIZE * 2,
  S = BASE_MARKER_SIZE,
}
export interface SliderProps {
  min: number;
  max: number;
  value: number;
  onChange: (value: number) => void;
  width?: number;
  size?: SliderSize;
}

const Slider = ({
  min,
  max,
  onChange,
  width = 300,
  size = BASE_MARKER_SIZE,
}: SliderProps) => {
  const translationX = useSharedValue(0);

  const onGestureEvent = useAnimatedGestureHandler<
    PanGestureHandlerGestureEvent,
    { startX: typeof translationX.value }
  >({
    onStart: (_, ctx) => {
      ctx.startX = translationX.value;
    },
    onActive: (event, ctx) => {
      translationX.value = interpolate(
        ctx.startX + event.translationX,
        [0, width - size],
        [0, width - size],
        Extrapolate.CLAMP,
      );
    },
    onEnd: () => {
      const newValue = interpolate(
        translationX.value,
        [0, width - size], // subtract marker width
        [min, max],
        Extrapolate.CLAMP,
      );
      onChange(newValue);
      translationX.value = withSpring(
        interpolate(
          newValue,
          [min, max],
          [0, width - size], // subtract marker width
          Extrapolate.CLAMP,
        ),
        { overshootClamping: true },
      );
    },
  });

  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: translationX.value }],
    };
  });
  const trackStyle = useAnimatedStyle(() => {
    const trackWidth = interpolate(
      translationX.value,
      [0, width],
      [size, width + size / 2],
      Extrapolate.CLAMP,
    );

    return {
      width: trackWidth,
    };
  });

  return (
    <SliderContainer style={{ width }} size={size}>
      <Tracker style={trackStyle} size={size} />
      <PanGestureHandler onGestureEvent={onGestureEvent}>
        <Marker style={[animatedStyle]} size={size}>
          <MarkerCircle size={size} />
        </Marker>
      </PanGestureHandler>
    </SliderContainer>
  );
};

type SliderCommonProps = {
  size?: SliderSize;
};
const SliderContainer = styled.View<SliderCommonProps>`
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  background-color: ${({ theme }) => theme.ds.colors.grayDark};
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;
  margin-vertical: 10px;
  overflow: visible;
  justify-content: center;
`;

const Tracker = styled(Animated.View)<SliderCommonProps>`
  position: absolute;
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;
  background-color: ${({ theme }) => theme.ds.colors.secondary};
`;
const Marker = styled(Animated.View)<SliderCommonProps>`
  width: ${({ size = SliderSize.S }) => size}px;
  height: ${({ size = SliderSize.S }) => size}px;
  border-radius: ${({ size = SliderSize.S }) => size / 2}px;
  background-color: ${({ theme }) => theme.ds.colors.light};
  align-items: center;
  justify-content: center;
  shadow-color: ${({ theme }) => theme.ds.colors.secondary};
  shadow-opacity: 0.2;
  shadow-radius: 10px;
  elevation: ${3};
  shadow-offset: {
    width: 0;
    height: 3px;
  }
`;
const MarkerCircle = styled.View<SliderCommonProps>`
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  width:  ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;
  background-color: ${({ theme }) => theme.ds.colors.secondary}};
`;

export default Slider;
