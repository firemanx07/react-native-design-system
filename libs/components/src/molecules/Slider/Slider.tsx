import { styled } from '@proxym/themes';
import React, { useCallback, useMemo, useState } from 'react';
import {
  PanGestureHandler,
  PanGestureHandlerGestureEvent,
} from 'react-native-gesture-handler';
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';

import { Badge, BadgeSize, BadgeVariant } from '../../atoms/Badge';
import { BaseText } from '../../atoms/BaseText';
import { formatTo3Digits } from '../../helpers/format-utils';

const BASE_MARKER_SIZE = 15;
export enum SliderSize {
  XL = BASE_MARKER_SIZE * 4,
  L = BASE_MARKER_SIZE * 3,
  M = BASE_MARKER_SIZE * 2,
  S = BASE_MARKER_SIZE,
}
export interface SliderProps {
  initialValue: number;
  label?: string;
  max: number;
  min: number;
  onChange: (value: number) => void;
  size?: SliderSize;
  width?: number;
}

const Slider = ({
  min,
  max,
  initialValue = min,
  onChange,
  width = 300,
  size = BASE_MARKER_SIZE,
  label = '',
}: SliderProps) => {
  const translationX = useSharedValue(
    interpolate(initialValue, [min, max], [0, width - size], Extrapolate.CLAMP),
  );

  const [currentValue, setCurrentValue] = useState(initialValue);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const onGestureEvent = useCallback(
    useAnimatedGestureHandler<
      PanGestureHandlerGestureEvent,
      { startX: typeof translationX.value }
    >({
      onStart: (_, ctx) => {
        ctx.startX = translationX.value;
      },
      onActive: (event, ctx) => {
        translationX.value = interpolate(
          ctx.startX + event.translationX,
          [0, width - size],
          [0, width - size],
          Extrapolate.CLAMP,
        );
        const newValue = interpolate(
          translationX.value,
          [0, width - size], // subtract marker width
          [min, max],
          Extrapolate.CLAMP,
        );
        Math.round(newValue) !== currentValue &&
          setCurrentValue(Math.round(newValue));
      },
      onEnd: () => {
        const newValue = interpolate(
          translationX.value,
          [0, width - size], // subtract marker width
          [min, max],
          Extrapolate.CLAMP,
        );
        onChange(Math.round(newValue));
        translationX.value = withSpring(
          interpolate(
            newValue,
            [min, max],
            [0, width - size], // subtract marker width
            Extrapolate.CLAMP,
          ),
          { overshootClamping: true },
        );
      },
    }),
    [currentValue, max, min, onChange, size, translationX.value, width],
  );
  const memoizedGestureHandler = useMemo(
    () => onGestureEvent,
    [onGestureEvent],
  );
  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: translationX.value }],
    };
  });
  const animatedBadgeStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: translationX.value - width / 10 + size / 2 }],
    };
  });
  const trackStyle = useAnimatedStyle(() => {
    const trackWidth = interpolate(
      translationX.value,
      [0, width],
      [size, width + size / 2],
      Extrapolate.CLAMP,
    );

    return {
      width: trackWidth,
    };
  });

  return (
    <RangeSelectorContainer>
      <PanGestureHandler onGestureEvent={memoizedGestureHandler}>
        <BadgeContainer style={animatedBadgeStyle}>
          <ArrowBadge
            variant={BadgeVariant.ArrowDown}
            size={BadgeSize.Regular}
            children={[formatTo3Digits(currentValue), label].join(' ')}
          />
        </BadgeContainer>
      </PanGestureHandler>
      <RangeTextContainer>
        <BaseText>{[formatTo3Digits(min), label].join(' ')}</BaseText>
        <BaseText>{[formatTo3Digits(max), label].join(' ')}</BaseText>
      </RangeTextContainer>
      <SliderContainer style={{ width }} size={size}>
        <Tracker style={trackStyle} size={size} />
        <PanGestureHandler onGestureEvent={memoizedGestureHandler}>
          <Marker style={[animatedStyle]} size={size}>
            <MarkerCircle size={size} />
          </Marker>
        </PanGestureHandler>
      </SliderContainer>
    </RangeSelectorContainer>
  );
};

type SliderCommonProps = {
  size?: SliderSize;
};
const SliderContainer = styled.View<SliderCommonProps>`
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  background-color: ${({ theme }) => theme.ds.colors.grayDark};
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;

  overflow: visible;
  justify-content: center;
`;

const Tracker = styled(Animated.View)<SliderCommonProps>`
  position: absolute;
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;
  background-color: ${({ theme }) => theme.ds.colors.secondary};
`;
const Marker = styled(Animated.View)<SliderCommonProps>`
  width: ${({ size = SliderSize.S }) => size}px;
  height: ${({ size = SliderSize.S }) => size}px;
  border-radius: ${({ size = SliderSize.S }) => size / 2}px;
  background-color: ${({ theme }) => theme.ds.colors.light};
  align-items: center;
  justify-content: center;
  shadow-color: ${({ theme }) => theme.ds.colors.secondary};
  shadow-opacity: 0.2;
  shadow-radius: 10px;
  elevation: ${3};
  shadow-offset: {
    width: 0;
    height: 3px;
  }
`;
const ArrowBadge = styled(Badge)`
  background-color: ${({ theme }) => theme.ds.colors.secondary}};
`;
const MarkerCircle = styled.View<SliderCommonProps>`
  height: ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  width:  ${({ size = SliderSize.S }) => (size * 2) / 3}px;
  border-radius: ${({ size = SliderSize.S }) => (size * 2) / 6}px;
  background-color: ${({ theme }) => theme.ds.colors.secondary}};
`;
const RangeSelectorContainer = styled.View``;
const RangeTextContainer = styled.View`
  flex-direction: row;
  justify-content: space-between;
  margin-vertical: 4px;
`;
const BadgeContainer = styled(Animated.View)`
  flex-direction: row;
  align-items: center;
  justify-content: center;
  max-width: 20%;
`;

export default Slider;
