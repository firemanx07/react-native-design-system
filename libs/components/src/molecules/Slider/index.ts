export { default as Slider, SliderSize } from './Slider';
