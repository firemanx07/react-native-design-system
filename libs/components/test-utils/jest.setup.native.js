import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();
require('react-native-gesture-handler/jestSetup.js');

console.warn = () => {
  //this is mock
};

Animated['timing'] = (value, config) => ({
  start: callback => {
    value.setValue(config.toValue);
    callback && callback({ finished: true });
  },
  stop: jest.fn(),
});
