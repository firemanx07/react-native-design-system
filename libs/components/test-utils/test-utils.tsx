import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
