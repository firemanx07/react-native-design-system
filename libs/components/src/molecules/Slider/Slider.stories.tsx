import { useNamespacedTheme } from '@proxym/themes';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen, StorybookSection } from '../../storybook';
import { default as SliderBase, SliderProps, SliderSize } from './Slider';

const Story: ComponentMeta<typeof SliderBase> = {
  component: SliderBase,
  title: 'Molecules/Slider',
  argTypes: {
    size: {
      control: { type: 'select' },
      options: Object.values(SliderSize).filter(
        item => typeof item === 'number',
      ),
    },
  },
};
export default Story;

const noop = () => {
  // This is a mock
};

const Template: ComponentStory<typeof SliderBase> = (args: SliderProps) => {
  const { colors } = useNamespacedTheme();
  return (
    <StorybookScreen
      title="Slider:"
      style={{ backgroundColor: colors.graySoft }}
    >
      <StorybookSection width={(args.width ?? 370) + 30}>
        <SliderBase {...args} />
      </StorybookSection>
    </StorybookScreen>
  );
};

const parameters = {
  controls: {
    include: [
      'label',
      'width',
      'size',
      'disabled',
      'min',
      'max',
      'step',
      'initialValue',
    ],
  },
};

const defaultArgs: SliderProps = {
  initialValue: 25,
  label: 'DT',
  min: 0,
  max: 100,
  onChange: noop,
  width: 300,
};

export const Slider = Template.bind({});
Slider.parameters = parameters;
Slider.args = {
  ...defaultArgs,
};
