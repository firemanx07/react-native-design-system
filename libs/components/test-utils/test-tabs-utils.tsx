import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
