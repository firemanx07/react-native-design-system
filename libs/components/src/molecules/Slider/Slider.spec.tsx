import { testAllStories } from '../../../test-utils';
import * as stories from './Slider.stories';

describe('Slider tests', () => {
  testAllStories(stories);
});
