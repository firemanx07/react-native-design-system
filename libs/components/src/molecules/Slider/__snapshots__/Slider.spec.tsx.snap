// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider tests Slider snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
      {
        "backgroundColor": "#DEDFE0",
      },
    ]
  }
>
  <Text
    fontFamily="Rubik-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#2C2C2C",
          "fontFamily": "Rubik-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Slider:
  </Text>
  <View
    style={
      [
        {
          "borderColor": "#2C2C2C",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 330,
        },
      ]
    }
    width={330}
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <View
        collapsable={false}
        handlerTag={1}
        handlerType="PanGestureHandler"
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "maxWidth": "20%",
            },
            {
              "transform": [
                {
                  "translateX": NaN,
                },
              ],
            },
          ]
        }
      >
        <View
          borderRadius={7}
          size={24}
          style={
            [
              {
                "alignItems": "center",
                "alignSelf": "flex-start",
                "backgroundColor": "#4D37D9",
                "borderRadius": 7,
                "flexDirection": "row",
                "height": 24,
                "justifyContent": "center",
                "minWidth": 24,
                "paddingBottom": 4,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 4,
              },
              {
                "backgroundColor": "#49C4E9",
              },
            ]
          }
        >
          <Text
            color="#FFFFFF"
            fontFamily="Rubik-Bold"
            fontSize={14}
            lineHeight={20}
            style={
              [
                {
                  "color": "#2C2C2C",
                  "fontFamily": "Rubik-Bold",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
                {
                  "color": "#FFFFFF",
                },
              ]
            }
          >
            25 DT
          </Text>
          <View
            style={
              [
                {
                  "backgroundColor": "#4D37D9",
                  "borderRadius": 2,
                  "bottom": -4,
                  "height": 10,
                  "position": "absolute",
                  "transform": [
                    {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 10,
                  "zIndex": -1,
                },
                {
                  "backgroundColor": "#49C4E9",
                },
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 4,
            },
          ]
        }
      >
        <Text
          fontFamily="Rubik-Regular"
          fontSize={16}
          lineHeight={20}
          style={
            [
              {
                "color": "#2C2C2C",
                "fontFamily": "Rubik-Regular",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          0 DT
        </Text>
        <Text
          fontFamily="Rubik-Regular"
          fontSize={16}
          lineHeight={20}
          style={
            [
              {
                "color": "#2C2C2C",
                "fontFamily": "Rubik-Regular",
                "fontSize": 16,
                "lineHeight": 20,
              },
            ]
          }
        >
          100 DT
        </Text>
      </View>
      <View
        size={15}
        style={
          [
            {
              "backgroundColor": "#C7C7D2",
              "borderRadius": 5,
              "height": 10,
              "justifyContent": "center",
              "overflow": "visible",
            },
            {
              "width": 300,
            },
          ]
        }
      >
        <View
          size={15}
          style={
            [
              {
                "backgroundColor": "#49C4E9",
                "borderRadius": 5,
                "height": 10,
                "position": "absolute",
              },
              {
                "width": undefined,
              },
            ]
          }
        />
        <View
          collapsable={false}
          handlerTag={2}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          size={15}
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 7.5,
                "elevation": 3,
                "height": 15,
                "justifyContent": "center",
                "shadowColor": "#49C4E9",
                "shadowOpacity": 0.2,
                "shadowRadius": 10,
                "width": 15,
              },
              {
                "transform": [
                  {
                    "translateX": undefined,
                  },
                ],
              },
            ]
          }
        >
          <View
            size={15}
            style={
              [
                {
                  "backgroundColor": "#49C4E9",
                  "borderRadius": 5,
                  "height": 10,
                  "width": 10,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
